### PR TITLE
armv7-a/r: modify regs index to improve exception performance

### PR DESCRIPTION
**Pull request by zyfeier**
## Summary

Modify arm regs index to improve exception handler performance.

## Impact

Code Size:
Before:

```
  text	   data	    bss	    dec	    hex	filename
 110164	    224	  24940	 135328	  210a0	nuttx
```

After:

```
  text	   data	    bss	    dec	    hex	filename
 110084	    224	  24940	 135248	  21050	nuttx
```

Performance:
sem_post -> sem_wait (cycle count):
before:
`358`

after:
`350`

## Testing
sabre-6quad:nsh

### CODE REVIEW

File: arch/arm/src/armv7-a/arm_vectors.S [hunk 1]
```diff
@@ -698,62 +652,50 @@ arm_vectorfiq:
...
 
 	mov		fp, #0				/* Init frame pointer */
 	mov		r0, sp				/* Get r0=xcp */
 
+	mov		r4, sp				/* Save the SP in a preserved register */
+
 #if CONFIG_ARCH_INTERRUPTSTACK > 7
 	/* Call arm_decodefiq() on the interrupt stack */
 
 	setfiqstack	r1, r4				/* SP = interrupt stack top */
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv7-a/arm_vectors.S:688]:
```suggestion
	setfiqstack	r1, r3				/* SP = interrupt stack top */
```

File: arch/arm/src/armv7-a/arm_vectors.S [hunk 2]
```diff
@@ -768,13 +710,14 @@ arm_vectorfiq:
 
 	/* Life is simple when everything is FIQ mode */
 
-	mov		r14, r0				/* (FIQ) r14=Register storage area */
+	ldmia		r0, {r13, r14}^			/* Restore user mode r13 and r14 */
+	add		r14, r0, #8
 	ldmia		r14!, {r0-r7}			/* Restore common r0-r7 */
-	ldmia		r14, {r8-r14}^			/* Restore user mode r8-r14 */
+	ldmia		r14, {r8-r12}^			/* Restore user mode r8-r12 */
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv7-a/arm_vectors.S:716]:
```suggestion
	ldmia		r14!, {r8-r12}^			/* Restore user mode r8-r12 */
```
and remove line 717

Comment (xiaoxiang781216) [reply arch/arm/src/armv7-a/arm_vectors.S:716]: please apply the same change to armv7-r

**Comment by xiaoxiang781216** [comment]
@masayuki2009 could you review this change?

File: arch/arm/src/armv7-a/arm_vectors.S (hunk 2 above)
Comment (zyfeier) [reply arch/arm/src/armv7-a/arm_vectors.S:716]: From the 《ARM Architecture Reference Manual》 B9.3.6 LDM (User registers):
```
In a PL1 mode other than System mode, Load Multiple (User registers) loads multiple User mode registers from
consecutive memory locations using an address from a base register. The registers loaded cannot include the PC.
The processor reads the base register value normally, using the current mode to determine the correct Banked
version of the register. This instruction cannot writeback to the base register.
```

**Review by xiaoxiang781216 — APPROVED** [review]
LGTM.

**Review by masayuki2009 — APPROVED** [review]

**Pull request merged**